### DISCUSSION TITLE
Метрики в технологической платформе

### DISCUSSION THREAD

**Discussion by bratchikov** [Flexberry/rfcs#17] [Ideas]
Собираем идеи как должны быть устроены метрики, которые через технологию "приедут" в прикладные проекты.

**Comment by mao29** [comment]
https://www.app-metrics.io/

**Comment by mao29** [comment]
Задачу предлагаю разбить на несколько этапов:

1. Посмотреть какие дефолтные метрики для webApi предлагаются в AppMetrics. С учетом специфики OData-запросов придумать какие метрики можно собирать по аналогии таким образом, чтобы они были полезны. Основные вопросы:
   * механизм формирования имени метрики из строки запроса
   * каким образом прикладной разработчик может настраивать и/или переопределять сбор метрик
   * эндпоинт для отправки метрик с фронтенда
   * каким образом проект может добавить сбор своих собственных прикладных метрик
   * каким образом добавить метрики в приложения, которые не хостятся в веб-сервере
Данные вопросы предлагаю описать в RFC и обсудить заинтересованными лицами.
2. Реализовать пакет со сбором метрик для ОДатаСервиса, добавить его использование в https://github.com/Flexberry/CodeGenSample.AspNetCorePlusEmber. Метрики пусть выставляются на некотором endpoint-е. Разработка экспортеров метрик - отдельная задача как мне кажется
3. Реализовать аддон для сбора метрик с фронтенда. В составе видится сервис запуска и отсечки времени какого-либо действия с отправкой значения на бекенд, миксины для добавления к роутам, доработка блюпринтов для генерации роутов, возможность включать/выключать сбор метрик для всего приложения и для конкретных роутов
4. Доработка генераторов бека и фронта, добавление флага UseMetrics в fgd, возможно также выбор экспортера (дефолтный - просто эндпоинт)
5. Документирование

**Comment by mao29** [comment]
По пункту 1 AppMetrics уже тоже считаются устаревшими. Теперь стандартом является OpenTelemetry, нужно посмотреть как его внедрить в технологию
